### PR TITLE
session as password

### PR DESCRIPTION
**Pull request by williamcroberts**
Things to Consider:
- I am not married to the option values, @joshuagl is going to start vetting these option strings and getting a table put together on the wiki (these making a ticket for this). When we agree on his table and choices, we can implement it going forward for 4.0, and then this will hopefully freeze command line options at that point.
- My patch granularity sucks. I should have had 1 for adding auth, and updating man pages and one for changing options. It's going to be a PITA to untangle, and considering that we're proposing a large option revamp for 4.0, I don't think it matters. The plus side is, -a is at consistent, and at least can be changed with sed.

Details:
Many tools need to specify N authorizations, where N > 0. Since a
single option cannot satisify this need,  and we want to
limit option explosion, move the session authorizaiotn logic
into the existing password logic. Now, sessions can be passed
via session file, be specifying "session:<path>" as the password.

Passwords have now been revamped to be clear that this is really one
of three authorization mechanisms and their respective options have
changed.

-a/--hierarchy is ALWAYS the hierarchy
??/--auth-XXX always auths to object XXX

The ?? tend to be -P most of the time.

The values chosen for the options might be subject to change in future
changes, but they are at least consistent for now.

Fixes: #904

Signed-off-by: William Roberts <william.c.roberts@intel.com>

### CODE REVIEW

File: lib/tpm2_auth_util.c [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        return false;
+    }
+
+    auth->sessionHandle = tpm2_session_get_handle(*session);
+
+    bool is_trial = tpm2_session_is_trial(*session);
+    if (is_trial) {
+        LOG_ERR("A trial session cannot be used to authenticate, "
+                "Please use an hmac or policy session");
+        tpm2_session_free(session);
```
Comment (martinezjavier) [inline lib/tpm2_auth_util.c]: Shouldn't this be `tpm2_session_free(*session) ?

File: lib/tpm2_auth_util.c [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+    /* str may or may not have the str: prefix */
+    bool is_str_prefix = !strncmp(password, STR_PREFIX, STR_PREFIX_LEN);
+    if (is_str_prefix) {
+        password += STR_PREFIX_LEN;
+    }
+
+    /*
+     * Per the man page:
+     * "a return value of size or more means that the output was  truncated."
```
Comment (martinezjavier) [inline lib/tpm2_auth_util.c]: There's an extra space before "truncated".

File: man/common/authorizations.md [hunk 3]
```diff
@@ -1,4 +1,13 @@
-# Password Formatting
+# Authorization Formatting
+
+Authorization for use of an object in tpm2.0 can come in 3 different forms:
```
Comment (martinezjavier) [inline man/common/authorizations.md]: We usually refer in the man page and docs as TPM 2.0 or TPM2.0, but never as tpm2.0.

File: man/common/authorizations.md [hunk 4]
```diff
@@ -1,4 +1,13 @@
...
+1. Password
+2. HMAC
+3. Sessions
+
+**NOTE:** Authorizations when not specified default to the **EMPTY PASSWORD**.
```
Comment (martinezjavier) [inline man/common/authorizations.md]: This sentence seems hard to read for me, maybe "When none specified, authorizations default to the **EMPTY PASSWORD**" or "Authorizations defaults to the **EMPTY PASSWORD** when not specified".

File: man/common/authorizations.md [hunk 5]
```diff
@@ -11,3 +20,14 @@ with special prefix values, they are:
...
          with the "hex:" prefix.
   * hex: - Used when specifying a password in hex string format.
+
+## HMAC
+
+HMAC tickets can be presented as hex escaped passwords.
+
+## Sessions
+
+When using a policy session to authorize use of an object, one prefixes the option argument
```
Comment (martinezjavier) [inline man/common/authorizations.md]: to authorize the use of an object.

File: man/tpm2_activatecredential.1.md [hunk 6]
```diff
@@ -33,7 +33,7 @@ These options control the object verification:
 
   * **-P**, **--password**=_PASSWORD_:
     Use _PASSWORD_ for providing an authorization value for the _KEY\_HANDLE_.
-    Passwords should follow the "password formatting standards, see section "Password Formatting".
+    Passwords should follow the "authorization formatting standards, see section "Authorization Formatting".
```
Comment (martinezjavier) [inline man/tpm2_activatecredential.1.md]: Missing enclosing `"` after standards.

Comment (martinezjavier) [reply man/tpm2_activatecredential.1.md]: Or probably just remove the opening `"` before authorization.

File: tools/tpm2_create.c [hunk 7]
```diff
@@ -164,10 +166,10 @@ int setup_alg()
     return 0;
 }
 
-int create(TSS2_SYS_CONTEXT *sapi_context)
```
Comment (martinezjavier) [inline tools/tpm2_create.c]: Probably I wouldn't mix cleanups like the create() return value in this commit, it makes harder to review. But I don't mind to have them either.

**Comment by martinezjavier** [comment]
@williamcroberts I very much like this, thanks for working on it. I just had some minor comments.

File: lib/tpm2_auth_util.c (hunk 1 above)
Comment (williamcroberts) [reply lib/tpm2_auth_util.c]: no tpm2_session_free() takes a double pointer and NULL's the pointer out.

File: man/common/authorizations.md (hunk 4 above)
Comment (williamcroberts) [reply man/common/authorizations.md]: "Authorizations default to the EMPTY PASSWORD when not specified"

File: man/tpm2_activatecredential.1.md (hunk 6 above)
Comment (williamcroberts) [reply man/tpm2_activatecredential.1.md]: Yeah I was removing the ", missed that one.

File: tools/tpm2_create.c (hunk 7 above)
Comment (williamcroberts) [reply tools/tpm2_create.c]: Yep my patch granularity was terrible. I just got in the habit of blindly collapsing my commits.

File: lib/tpm2_auth_util.c (hunk 1 above)
Comment (martinezjavier) [reply lib/tpm2_auth_util.c]: Ah, sorry for misreading that then.

File: tools/tpm2_evictcontrol.c [hunk 8]
```diff
@@ -134,17 +131,14 @@ static bool on_option(char key, char *value) {
 bool tpm2_tool_onstart(tpm2_options **opts) {
 
     const struct option topts[] = {
-      { "auth",                 required_argument, NULL, 'A' },
+      { "hierarchy",       required_argument, NULL, 'a' },
       { "handle",               required_argument, NULL, 'H' },
       { "persistent",           required_argument, NULL, 'p' },
-      { "pwda",                 required_argument, NULL, 'P' },
+      { "auth-hierarchy", required_argument, NULL, 'P' },
```
Comment (liuqun) [inline tools/tpm2_evictcontrol.c]: tools/tpm2_evictcontrol.c doesn't have `--auth-hierarchy-value` as CHANGELOG.md says

File: man/tpm2_evictcontrol.1.md [hunk 9]
```diff
@@ -44,10 +44,10 @@ be evicted.
 
     The persistent handle for the object handle specified via _HANDLE_.
 
-  * **-P**, **--pwda**=_AUTH\_PASSWORD_:
+  * **-P**, **--auth-hierarchy**=_AUTH\_HIERARCHY_\VALUE_:
```
Comment (liuqun) [inline man/tpm2_evictcontrol.1.md]: This is `--auth-hierarchy`. The previous document is wrong.

File: CHANGELOG.md [hunk 10]
```diff
@@ -1,5 +1,46 @@
...
+    respectively.
+  * tpm2_loadexternal: `-H` changes to `-a` for specifying hierarchy.
+  * tpm2_load: Option `--pwdp` changes to `--auth-parent`.
+  * tpm2_hmac: Option `--pwdk` changes to `--auth-key`.
+  * tpm2_hash: `-H` changes to `-a` for specifying hierarchy.
+  * tpm2_getmanufec: Options `--owner-passwd`, `--endorse-passwd`
+  * and `--ek-passwd`change to `--auth-owner`, `--auth-endorse`
+    and `--auth-ek` respectively.
+  * tpm2_evictcontrol: Option group `-A` and `--auth` changes to `-a` and `--hierarchy`
+    Option `--pwda` changes to `--auth-hierarchy-value`
```
[the remark this answers is not in the document]
Comment (liuqun) [reply CHANGELOG.md]: Maybe  `--auth-hierarchy=value`?

File: man/tpm2_evictcontrol.1.md (hunk 9 above)
Comment (liuqun) [reply man/tpm2_evictcontrol.1.md]:
```
_AUTH\_HIERARCHY_\VALUE_
```
should be
```
_AUTH\_HIERARCHY\_VALUE_
```

File: man/tpm2_evictcontrol.1.md [hunk 11]
```diff
@@ -17,7 +17,7 @@ be evicted.
 
 # OPTIONS
 
-  * **-A**, **--auth**=_AUTH_:
+  * **-a**, **--auth-heirarchy**=_AUTH\_HIERARCHY\_:
```
[the remark this answers is not in the document]
Comment (liuqun) [reply man/tpm2_evictcontrol.1.md]: It should be --heirarchy as PR https://github.com/tpm2-software/tpm2-tools/pull/965 fixed

[the remark this answers is not in the document]
Comment (dantpm) [reply man/tpm2_evictcontrol.1.md]: Please spell the command line option correctly.  It is --hierarchy (not heirarchy)

[the remark this answers is not in the document]
Comment (liuqun) [reply man/tpm2_evictcontrol.1.md]: PR #965 had fixed the spelling of `--hierarchy`.

**Pull request merged**